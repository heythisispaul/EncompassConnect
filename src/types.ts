/* eslint-disable camelcase */
/* eslint-disable no-undef */
/* eslint-disable no-unused-vars */

/**
 * @packageDocumentation
 * @module Interfaces
 */
export interface SortOrderContract {
  canonicalName: string;
  order: 'asc' | 'desc';
}

export interface PipeLineRequest {
  fields?: string[];
  sortOrder?: SortOrderContract[];
}

export interface LoanGuidsPipeLineContract extends PipeLineRequest {
  loanGuids: string[];
}

export interface PipeLineTerms {
  canonicalName: string;
  matchType: 'greaterThanOrEquals' | 'exact' | 'greaterThan' | 'isNotEmpty' | 'isEmpty' | 'lessThan' | 'lessThanOrEquals' | 'equals' | 'notEquals' | 'startsWith' | 'contains';
  value?: string | number | Date;
  precision?: 'exact'| 'day' | 'month' | 'year' | 'recurring';
}

export interface PipeLineFilter {
  operator: 'and' | 'or';
  terms: (PipeLineFilter | PipeLineTerms)[];
}

export interface FilterPipeLineContract extends PipeLineRequest {
  filter: PipeLineFilter;
}

export type PipeLineContract = LoanGuidsPipeLineContract | FilterPipeLineContract;

export interface BatchLoanUpdateContract {
  filter?: PipeLineFilter;
  loanGuids?: string[];
  loanData: any;
}

export interface LoanAssociateProperties {
  loanAssociateType: 'user' | 'group';
  id: string;
  name?: string;
  phone?: string;
  cellphone?: string;
  fax?: string;
  email?: string;
  roleId?: string;
  roleName?: string;
}

export interface UserInfoContract {
  viewEmailSignature?: boolean;
  start?: number;
  limit?: number | string;
  filter?: {
    groupId?: string | number;
    roleId?: string | number;
    personaId?: string | number;
    organizationId?: string | number;
    userName?: string;
  }
}

export interface Organization {
  entityId: string | number;
  entityType: string;
  entityName: string;
  entityUri: string;
}

export interface UserProfile {
  id: string;
  lastName: string;
  firstName: string;
  fullName: string;
  email: string;
  phone: string;
  userIndicators: string[],
  peerLoanAccess: string;
  lastLogin: string | Date;
  encompassVersion: string;
  personalStatusOnline: boolean;
  personas: any[];
  workingFolder?: string;
  organization?: Organization;
  subordinateLoanAccess?: string;
  comments?: string;
  ccSite?: any;
}

export interface LicenseInformation {
  state: string;
  enabled: boolean;
  license?: string;
  expirationDate?: string | Date;
  issueDate?: string | Date;
  startDate?: string | Date;
}

export interface CreateLoanContract {
  view?: 'entity' | 'id';
  loanTemplate?: string;
  loanFolder?: string;
  loan?: any;
}

export interface EncompassConnectInitOptions {
  clientId: string;
  APIsecret: string;
  instanceId: string;
  username?: string;
  password?: string;
  version?: number;
}

export interface InternalRequestOptions {
  isRetry?: boolean;
  isNotJson?: boolean;
  version?: number;
  useTruncatedBase?: boolean;
}

export interface AssignMilestoneOptions {
  loanGuid: string;
  milestone: string;
  userId: string;
}

export interface UpdateMilestoneOptions {
  loanGuid: string;
  milestone: string;
  options: any;
  action?: 'finish' | 'unfinish';
}

export interface LoanUpdateOptions {
  appendData: boolean | string;
  persistent: 'transient' | 'permanent';
  view: 'entity' | 'id';
  loanTemplate?: string;
}

export interface BatchUpdateStatus {
  status: string;
  lastModified: string;
}

export interface BatchUpdate {
  getRequestId: () => string;
  getUpdateStatus: () => Promise<BatchUpdateStatus>;
}

export interface UpdateLoanWithGenerateContract {
  standardFields?: {
    [key: string]: any;
  },
  customFields?: {
    [key: string]: any;
  }
}

export interface TokenIntrospection {
  active: boolean;
  scope: string;
  client_id: string;
  username: string;
  token_type: 'Bearer';
  exp: number;
  sub: string;
  encompass_instance_id: string;
  user_name: string;
  user_key: string;
  encompass_user: string;
  identity_type: 'Enterprise';
  encompass_instance_type: string;
  encompass_client_id: string;
  realm_name: string;
  bearer_token?: string;
}

export interface ListOfUsersOptions {
  viewEmailSignature?: boolean;
  groupId?: string | number;
  roleId?: string;
  personaId?: string | number;
  organizationId?: string | number;
  userName?: string;
  start?: number;
  limit?: number;
}
