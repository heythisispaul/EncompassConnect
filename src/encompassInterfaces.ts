/* eslint-disable no-undef */
/* eslint-disable no-unused-vars */
export interface sortOrderContract {
  canonicalName: string;
  order: 'asc' | 'desc';
}

export interface PipeLineRequest {
  fields?: string[];
  sortOrder?: sortOrderContract[];
}

export interface LoanGuidsPipeLineContract extends PipeLineRequest {
  loanGuids: string[];
}

export interface PipeLineTerms {
  canonicalName: string;
  matchType: 'greaterThanOrEquals' | 'exact' | 'greaterThan' | 'isNotEmpty' | 'isEmpty' | 'lessThan' | 'lessThanOrEquals' | 'equals' | 'notEquals' | 'startsWith' | 'contains';
  value?: string | number | Date;
  precision?: 'exact'| 'day' | 'month' | 'year' | 'recurring';
}

export interface PipeLineFilter {
  operator: 'and' | 'or';
  terms: PipeLineTerms[];
}

export interface FilterPipeLineContract extends PipeLineRequest {
  filter: PipeLineFilter;
}

export type PipeLineContract = LoanGuidsPipeLineContract | FilterPipeLineContract;

export interface BatchLoanUpdateContract {
  filter?: PipeLineFilter;
  loanGuids?: string[];
  loanData: any;
}

export interface LoanAssociateProperties {
  loanAssociateType: 'user' | 'group';
  id: string;
  name?: string;
  phone?: string;
  cellphone?: string;
  fax?: string;
  email?: string;
  roleId?: string;
  roleName?: string;
}

export interface UserInfoContract {
  viewEmailSignature?: boolean;
  start?: number;
  limit?: number | string;
  filter?: {
    groupId?: string | number;
    roleId?: string | number;
    personaId?: string | number;
    organizationId?: string | number;
    userName?: string;
  }
}

export interface Organization {
  entityId: string | number;
  entityType: string;
  entityName: string;
  entityUri: string;
}

export interface UserProfile {
  id: string;
  lastName: string;
  firstName: string;
  fullName: string;
  email: string;
  phone: string;
  userIndicators: string[],
  peerLoanAccess: string;
  lastLogin: string | Date;
  encompassVersion: string;
  personalStatusOnline: boolean;
  personas: any[];
  workingFolder?: string;
  organization?: Organization;
  subordinateLoanAccess?: string;
  comments?: string;
  ccSite?: any;
}

export interface LicenseInformation {
  state: string;
  enabled: boolean;
  license?: string;
  expirationDate?: string | Date;
  issueDate?: string | Date;
  startDate?: string | Date;
}

export interface CreateLoanContract {
  view?: 'entity' | 'id';
  loanTemplate?: string;
  loanFolder?: string;
  loan?: any;
}

export interface EncompassConnectConstructor {
  clientId: string;
  APIsecret: string;
  instanceId: string;
  username?: string;
  password?: string;
}

export interface InternalRequestOptions {
  isRetry?: boolean;
  isNotJson?: boolean;
}

export interface AssignMilestone {
  loanGuid: string;
  milestone: string;
  userId: string;
}

export interface UpdateMilestone {
  loanGuid: string;
  milestone: string;
  options: any;
  action?: 'finish' | 'unfinish';
}

export interface LoanUpdateOptions {
  appendData: boolean;
  persistent: 'transient' | 'permanent';
  view: 'entity' | 'id';
  loanTemplate?: string;
}
