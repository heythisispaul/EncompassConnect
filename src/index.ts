import EncompassConnect from './encompassConnect';

export type { Response, RequestInit } from 'node-fetch';
export type {
  AssignMilestoneOptions,
  BatchLoanUpdateContract,
  EncompassConnectInitOptions,
  PipeLineContract,
  UpdateMilestoneOptions,
  LoanUpdateOptions,
  BatchUpdateStatus,
  UpdateLoanWithGenerateContract,
  BatchUpdate,
  TokenIntrospection,
  ListOfUsersOptions,
} from './types';

export default EncompassConnect;
