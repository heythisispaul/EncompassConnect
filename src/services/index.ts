export { default as LoanService } from './loans';
export { default as MilestoneService } from './milestones';
export { default as SchemaService } from './schemas';
export { default as UserService } from './users';
