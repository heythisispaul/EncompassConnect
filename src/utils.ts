export const massageCustomFields = (customFields: any) => {
  if (!customFields) {
    return [];
  }
  const massaged = Object
    .entries(customFields).map(([fieldName, value]: [string, any]) => {
      const parsedNumber = parseFloat(value);
      const customFieldObject: any = {
        fieldName,
        stringValue: value && value.toString ? value.toString() : value,
      };
      if (!Number.isNaN(parsedNumber)) {
        customFieldObject.numericValue = parsedNumber;
      }
      return customFieldObject;
    });
  return massaged;
};

export const objectToURLString = (input: any, queryString?: boolean): string => {
  // @ts-ignore
  const result = input ? new URLSearchParams(input).toString() : '';
  return queryString ? `?${result}` : result;
};
