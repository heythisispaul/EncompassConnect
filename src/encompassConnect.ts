/* eslint-disable max-len */
/* eslint-disable camelcase */

/**
 * @packageDocumentation
 * @module EncompassConnectClass
 */
import fetch, {
  Response,
  RequestInit,
  HeadersInit,
} from 'node-fetch';
import {
  EncompassConnectInitOptions,
  InternalRequestOptions,
  PipeLineContract,
  BatchLoanUpdateContract,
  BatchUpdateStatus,
  BatchUpdate,
  TokenIntrospection,
} from './types';
import {
  LoanService,
  MilestoneService,
  SchemaService,
  UserService,
} from './services';
import { objectToURLString } from './utils';

class EncompassConnect {
  /**
   * The client ID of the Encompass instance to connect to.
   */
  #clientId: string;

  /**
   * The API key of the Encompass instance to connect to.
   */
  #APIsecret: string;

  /**
   * The Encompass instance ID of the instance to connect to.
   */
  #instanceId: string;

  /**
   * The password of the account to retrieve tokens with. Only stores the value provided to the constructor.
   */
  #password: string;

  /**
   * The bearer token that is used to authenticate each request to the Encompas API. The same token will be reused
   * until either a `401` is returned, or the value is overwritten with the `setToken()` or `getToken()` methods.
   */
  #token: string | null;

  /**
   * The username of the account to retrieve tokens with. Only stores the value provided to the constructor.
   */
  username: string;

  /**
   * The domain of the Encompass API.
   */
  base: string;

  loans: LoanService;

  milestones: MilestoneService;

  schemas: SchemaService;

  users: UserService;

  /**
   * @ignore
   */
  authBase: string;

  /**
   * The version of the Encompass API to send requests to. Defaults to `1`.
   */
  version: number;

  constructor({
    clientId,
    APIsecret,
    instanceId,
    username,
    password,
    version = 1,
  }: EncompassConnectInitOptions) {
    this.#clientId = clientId;
    this.#APIsecret = APIsecret;
    this.#instanceId = instanceId;
    this.#password = password || '';
    this.#token = '';
    this.username = username || '';
    this.version = version;
    this.base = 'https://api.elliemae.com/encompass';
    this.authBase = 'https://api.elliemae.com';
    this.loans = new LoanService(this);
    this.milestones = new MilestoneService(this);
    this.schemas = new SchemaService(this);
    this.users = new UserService(this);
  }

  /**
   * Replaces the `#token` property with the provided token value. The instance can be implicitly 'logged out' by setting this value to `null` (if it was not provided a username and password in the constructor).
   */
  setToken(token:string | null): void {
    this.#token = token;
  }

  /**
   * @ignore
   */
  private withTokenHeader(headers = {}): HeadersInit {
    return {
      ...headers,
      Authorization: `Bearer ${this.#token}`,
    };
  }

  /**
   * @ignore
   */
  async fetchWithRetry(
    path: string,
    options: RequestInit = {},
    customOptions: InternalRequestOptions = {},
  ): Promise<any> {
    const { isRetry, isNotJson, version } = customOptions;
    const shouldRetry = !isRetry && this.username && this.#password;
    const failedAuthError = new Error(
      `Token invalid. ${!isRetry && shouldRetry ? 'Will reattempt with new token' : 'Unable to get updated one.'}`,
    );
    try {
      if (!this.#token) {
        await this.getToken();
      }
      const url = `${this.base}/v${version || this.version}${path}`;
      const optionsWithToken: RequestInit = {
        ...options,
        headers: this.withTokenHeader(options.headers),
      };
      const response: Response = await fetch(url, optionsWithToken);
      if (response.status === 401) {
        throw failedAuthError;
      }
      if (!response.ok) {
        throw new Error(response.statusText);
      }
      return isNotJson ? response : await response.json();
    } catch (error) {
      if (shouldRetry && error === failedAuthError) {
        this.setToken(null);
        return this.fetchWithRetry(path, options, {
          ...customOptions,
          isRetry: true,
        });
      }
      throw error;
    }
  }

  /**
   * Exchanges the provided username and password for a bearer token and stores it to the `#token` property of the instance.
   * If no username and password are provided, it will fallback to the username and password values provided to the constructor.
   */
  async getToken(username?: string, password?: string): Promise<void> {
    const body: string = objectToURLString({
      grant_type: 'password',
      username: `${username || this.username}@encompass:${this.#instanceId}`,
      password: password || this.#password,
      client_id: this.#clientId,
      client_secret: this.#APIsecret,
    });

    const requestOptions: RequestInit = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body,
      redirect: 'follow',
    };

    const response = await fetch(`${this.authBase}/oauth2/v1/token`, requestOptions);
    const { access_token } = await response.json();
    this.setToken(access_token);
  }

  /**
   * Calls the token introspection API with the provided token. If a token is not provided, it will introspect the token stored to the `#token` property of the instance as a fallback.
   * If the introspection returns a valid token, it will return the response body of the request, if the token is invalid, returns `null`.
   */
  async introspectToken(token?: string): Promise<TokenIntrospection | null> {
    const body: string = objectToURLString({
      token: token || this.#token,
    });
    const Authorization: string = `Basic ${Buffer.from(`${this.#clientId}:${this.#APIsecret}`).toString('base64')}`;
    const requestOptions: RequestInit = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        Authorization,
      },
      body,
    };
    try {
      const response = await fetch(`${this.authBase}/oauth2/v1/token/introspection`, requestOptions);
      if (response.ok) {
        const tokenData: TokenIntrospection = await response.json();
        return tokenData;
      }
      return null;
    } catch (error) {
      return null;
    }
  }

  /**
   * Returns the result of the 'Get Canonical Names' endpoint.
   */
  async getCanonicalNames(): Promise<any> {
    const canonicalNames = await this.fetchWithRetry('/loanPipeline/fieldDefinitions');
    return canonicalNames;
  }

  /**
    *  Generate a pipeline view by calling the `viewPipeline()` method. This method has one required argument, a `PipeLineContract`, and can optionally take a limit value as the second argument:
    *
    *  ```typescript
    *  // a pipelineContract expects either a loanGuids array, or a filter object:
    *  const commonFilterValues = {
    *    sortOrder: [
    *      {
    *        canonicalName: 'Loan.LastModified',
    *        order: 'desc'
    *      }
    *    ],
    *    fields: [
    *      "Loan.LoanAmount",
    *      "Fields.4002"
    *    ],
    *  };
    *
    *  const pipelineWithGuids: LoanGuidsPipeLineContract = {
    *    ...commonFilterValues,
    *    loanGuids: [
    *      'some-loan-guid-1',
    *      'some-loan-guid-2',
    *    ],
    *  };
    *
    *  const pipelineWithFilter: FilterPipeLineContract = {
    *    ...commonFilterValues,
    *    filter: {
    *      operator: 'and',
    *      terms: [
    *        {
    *          canonicalName: "Loan.LastModified",
    *          matchType: "greaterThanOrEquals",
    *          value: new Date()
    *        },
    *        {
    *          canonicalName: "Loan.LoanFolder",
    *          matchType: "exact",
    *          value: "My Pipeline"
    *        }
    *      ]
    *    },
    *  };
    *
    *  const pipelineDataFromGuids = await encompass.viewPipeline(pipelineWithGuids);
    *
    *  // or with the other contract, and a limit of the first 50 results:
    *  const pipelineDataFromFilter = await encompass.viewPipeline(pipelineWithFilter, 50);
    *  ```
    */
  async viewPipeline(options: PipeLineContract, limit?: number) {
    const uri = `/loanPipeline${limit ? `?limit=${limit}` : ''}`;
    const pipeLineData = await this.fetchWithRetry(uri, {
      method: 'POST',
      headers: this.withTokenHeader(),
      body: JSON.stringify(options),
    });
    return pipeLineData;
  }

  /**
    *  The batch update API allows your to apply the same loan data to multiple loans and can be invoked with `batchLoanUpdate()` method.
    *
    *  This method returns an object that with it's own functionality to check the status of batch update, or to get the request ID if needed.
    *
    *  Just like viewing a pipeline, either a filter or an array of lan GUIDs can be provided.
    *  ```typescript
    *  const updateData: BatchLoanUpdateContract = {
    *    loanGuids: [
    *      // array of loan GUIDs to update
    *    ],
    *    loanData: {
    *      // contract of the loan data to apply to each loan
    *    },
    *  };
    *
    *  const exampleBatchUpdate: BatchUpdate = await encompass.batchLoanUpdate(updateData);
    *
    *  // the return value can be used to check the status:
    *  const latestStatus: BatchUpdateStatus = await exampleBatchUpdate.getUpdateStatus();
    *  console.log(latestStatus.status) // 'done' or 'error'
    *
    *  // or if needed you can get the request ID itself:
    *  const exampleBatchUpdateId: string = exampleBatchUpdate.getRequestId();
    *  ```
    */
  async batchLoanUpdate(options: BatchLoanUpdateContract): Promise<BatchUpdate> {
    const response: any = await this.fetchWithRetry('/loanBatch/updateRequests', {
      method: 'POST',
      body: JSON.stringify(options),
    }, { isNotJson: true });
    const requestId: string = response && response.headers
      ? response.headers.get('location').split('/').reverse()[0]
      : null;
    return {
      getRequestId: () => requestId,
      getUpdateStatus: async () => {
        const url = `/loanBatch/updateRequests/${requestId}`;
        const status: BatchUpdateStatus = await this.fetchWithRetry(url);
        return status;
      },
    };
  }

  /**
    * If an API is not available through an explicit method in this class, the `request()` method is a wrapper around fetch that allows you to request any Encompass API endpoint.
    *
    * It takes in the same arguments as any [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) API, with the exception that the first argument is appended as a path to the Encompass API domain.
    * ```typescript
    * // hitting the Get Custom Fields API:
    * const customFieldsResponse: Response = await encompass.request('/settings/loan/customFields');
    * const data = await customFieldsResponse.json();
    * console.log(data);
    *
    * // or update a contact:
    * const options: RequestInit = {
    *   method: 'POST',
    *   body: {
    *     firstname: 'contact first name',
    *     lastname: 'contact last name',
    *   },
    * };
    * await encompass.request('/businessContacts/<some-contact-id>', options);
    * ```
    */
  async request(url: string, options: RequestInit): Promise<Response> {
    const response = await this.fetchWithRetry(url, options, { isNotJson: true });
    return response;
  }
}

export default EncompassConnect;
