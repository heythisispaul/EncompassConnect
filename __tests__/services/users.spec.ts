import { fetch } from '../__utils__/mockFunctions';
import { matchesFetchSnapshot } from '../__utils__/commonTests';
import { ListOfUsersOptions } from '../../src/types';

jest.mock('node-fetch');

describe('UserService', () => {
  beforeEach(() => {
    fetch.mockReset();
  });

  describe('getProfile', () => {
    it('makes the API call with provided LOS Id in the url', async () => {
      matchesFetchSnapshot('users.getProfile', '<TEST_LOS_ID>');
    });

    it('makes the API call with "me" in the url if no argument is provided', async () => {
      matchesFetchSnapshot('users.getProfile');
    });
  });

  describe('getList', () => {
    it('makes the request to the correct URL', () => {
      matchesFetchSnapshot('users.getList');
    });

    it('appends the filter object as a query string', () => {
      const options: ListOfUsersOptions = {
        groupId: 1,
        limit: 50,
        userName: 'Mr. Peanutbutter',
      };
      matchesFetchSnapshot('users.getList', options);
    });
  });
});
