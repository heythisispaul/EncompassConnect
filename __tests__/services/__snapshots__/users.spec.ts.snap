// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserService getList appends the filter object as a query string 1`] = `
Array [
  "https://api.elliemae.com/encompass/v1/company/users?",
  Object {
    "headers": Object {
      "Authorization": "Bearer <MOCK TOKEN VALUE>",
    },
  },
]
`;

exports[`UserService getList appends the filter object as a query string 2`] = `
Array [
  "https://api.elliemae.com/encompass/v1/company/users?groupId=1&limit=50&userName=Mr.+Peanutbutter",
  Object {
    "headers": Object {
      "Authorization": "Bearer <MOCK TOKEN VALUE>",
    },
  },
]
`;

exports[`UserService getProfile makes the API call with "me" in the url if no argument is provided 1`] = `
Array [
  "https://api.elliemae.com/encompass/v1/company/users/<TEST_LOS_ID>",
  Object {
    "headers": Object {
      "Authorization": "Bearer <MOCK TOKEN VALUE>",
    },
  },
]
`;

exports[`UserService getProfile makes the API call with "me" in the url if no argument is provided 2`] = `
Array [
  "https://api.elliemae.com/encompass/v1/company/users/me",
  Object {
    "headers": Object {
      "Authorization": "Bearer <MOCK TOKEN VALUE>",
    },
  },
]
`;
